src/leveling.ts

```diff
@@ -0,0 +1,213 @@
+import { EmbedBuilder } from 'discord.js';
+import { getDB, getDBList } from '@/db';
+import { User } from '@/models/users';
+import { sendEventLogMessage } from '@/util';
+import type { GuildMember, Message } from 'discord.js';
+
+export async function handleLeveling(message: Message): Promise<void> {
+	const levelingChannelBlacklist = getDBList('leveling.channels-blacklist');
+	if (levelingChannelBlacklist.includes(message.channelId)) {
+		return;
+	}
+
+	const supporterRoleID = getDB().get('roles.supporter');
+	const supporterRole = supporterRoleID && (await message.guild?.roles.fetch(supporterRoleID));
+	if (!supporterRole) {
+		console.log('Missing supporter role!');
+		return;
+	}
+
+	const trustedRoleID = getDB().get('roles.trusted');
+	const trustedRole = trustedRoleID && (await message.guild?.roles.fetch(trustedRoleID));
+	if (!trustedRole) {
+		console.log('Missing trusted role!');
+		return;
+	}
+
+	const xpRequiredForTrusted = parseInt(getDB().get('leveling.xp-required-for-trusted') ?? '');
+	if (isNaN(xpRequiredForTrusted)) {
+		console.log('Missing amount of XP required for trusted role!');
+		return;
+	}
+
+	const timeRequiredForTrusted = parseInt(getDB().get('leveling.days-required-for-trusted') ?? '') * 24 * 60 * 60 * 1000;
+	if (isNaN(timeRequiredForTrusted)) {
+		console.log('Missing number of days required for trusted role!');
+		return;
+	}
+
+	let messageTimeout = parseInt(getDB().get('leveling.message-timeout-seconds') ?? '') * 1000;
+	if (isNaN(messageTimeout)) {
+		console.log('Missing leveling message timeout! Defaulting to 1 minute.');
+		messageTimeout = 60 * 1000;
+	}
+
+	let supporterXPMultiplier = parseInt(getDB().get('leveling.supporter-xp-multiplier') ?? '');
+	if (isNaN(supporterXPMultiplier)) {
+		console.log('Missing supporter XP multiplier! Supporters will not earn extra XP.');
+		supporterXPMultiplier = 1;
+	}
+
+	let user = await User.findOne({
+		where: {
+			user_id: message.author.id
+		}
+	});
+
+	if (!user) {
+		user = await User.create({
+			user_id: message.author.id
+		});
+	}
+
+	const joinDate = message.member?.joinedAt;
+	if (!joinDate) {
+		// * User has left, no point in tracking their XP
+		return;
+	}
+
+	if (!user.trusted_time_start_date) {
+		// * User has not used the leveling system yet, set their start date to their join date so they don't have to wait
+		user.trusted_time_start_date = joinDate;
```

Comment:
Yes, but we still need to check this in case the user was already created without a start date.

Reply:
If we use `DataTypes.NOW` as the default value in the model, I think you can remove this

Reply:
While that would technically work, having the start date be the current date is not intended. Since the User model is also used by the matchmaking threads functionality to check if the user has already received their one-time notification, it is theoretically possible (although very unlikely due to how the timing works) for a user to be created outside of the leveling function. In this situation, it would not be correct to create the user with the start date being now.